File: Slowsort.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class Slowsort {
	public static final int LENGTH = 1000;
	public static final int MAX = 1000;

	public static void main(String[] args) {
		Random rand = new Random();
		List<Integer> toSort;
//		print(toSort);
		double ratio = 0;
		double buff = 1;
		for (int j = 50; j < LENGTH; j+=50) {
			toSort = createList(rand, j);
			long t = System.currentTimeMillis();
			slowsort(toSort, 0, toSort.size() - 1);
			double elapsed = (System.currentTimeMillis() - t) / 1000.0;
			double ops = elapsed*1000;
			String unit = "Seconds";
			if (j != 2 && buff != 0) {
				buff = buff /1.0;
				ratio = ops / buff;
			}
			if (elapsed > 60) {
				elapsed = elapsed / 60;
				unit = "Minutes";
			}
			if (elapsed > 60) {
				elapsed = elapsed / 60;
				unit = "Hours";
			}
//			print(toSort);
//			if (j % 50 == 0) {
//				System.out.printf("%f %s for %d numbers\n", elapsed / 1.0, unit, j);
//			}
			System.out.printf("%f %s for %d numbers; Ratio = %f\n", elapsed / 1.0, unit, j, ratio);
//			toSort = addOne(rand, toSort);
			if (ops == 0) {
				buff = 1;
			} else {
				buff = ops;
			}
		}
	}

	public static List<Integer> createList(Random rand, int length) {
		List<Integer> numList = new ArrayList<Integer>();
		for (int i = 0; i < length; i++) {
			numList.add(rand.nextInt(MAX));
		}
		return numList;
	}

	public static List<Integer> addOne(Random rand, List<Integer> list) {
		list.add(rand.nextInt(MAX));
		Collections.shuffle(list);
		return list;
	}

	public static void slowsort(List<Integer> list, int i, int j) {
		if (i >= j) {
			return;
		}
		int m = (i + j) / 2;
		slowsort(list, i, m);
		slowsort(list, m + 1, j);
		if (list.get(j) < list.get(m)) {
			int temp = list.get(j);
			list.set(j, list.get(m));
			list.set(m, temp);
		}
		slowsort(list, i, j - 1);
	}

	private static void print(List<Integer> numList) {
		for (int x : numList) {
			System.out.printf("%d ", x);
		}
		System.out.println();
	}

}

File: Slowsort.py
import random
from time import perf_counter
import sys

MAX = LENGTH = 1000
STEP = 50


def make_list(num_list=[], to_add=LENGTH):
    for i in range(0, to_add):
        num_list.append(random.randint(0, MAX))
        # print(len(numList))
    return num_list


def slow_sort(A, i, j):
    if i >= j:
        return
    m = int((i + j) / 2)
    slow_sort(A, i, m)
    slow_sort(A, m + 1, j)
    if A[j] < A[m]:
        temp = A[j]
        A[j] = A[m]
        A[m] = temp
    slow_sort(A, i, j - 1)


def run_sort():
    buff = 1
    ratio = 0.0
    nums = []
    for i in range(STEP, LENGTH+STEP, STEP):
        if len(nums) == 0:
            nums = make_list(to_add=i)
        else:
            random.shuffle(nums)
            nums = make_list(num_list=nums, to_add=STEP)
        t1_start = perf_counter()
        slow_sort(nums, 0, len(nums) - 1)
        t1_stop = perf_counter()
        elapsed = ops = t1_stop - t1_start
        if buff != 0:
            buff = buff / 1.0
            ratio = ops/buff
            unit = "Seconds"
        if elapsed > 60.0:
            elapsed = elapsed / 60
            unit = "Minutes"
        if elapsed > 60.0:
            elapsed = elapsed / 60
            unit = "Hours"
        print(f'{elapsed/1.0} {unit} for {i} numbers; Ratio = {ratio}\n')
        if ops == 0:
            buff = 1.0
        else:
            buff = ops


if __name__ == '__main__':
    sys.setrecursionlimit(1500)
    run_sort()
